File: src/main/java/com/ottofernan/librarycrud/domain/models/Book.java
package com.ottofernan.librarycrud.domain.models;


import com.ottofernan.librarycrud.domain.dtos.BookDTO;
import com.ottofernan.librarycrud.domain.dtos.VisitorDTO;

import javax.persistence.*;
import java.util.HashSet;
import java.util.Set;
import java.util.stream.Collectors;
import javax.persistence.Table;

import static com.ottofernan.librarycrud.domain.dtos.VisitorDTOKt.toModel;
import static com.ottofernan.librarycrud.domain.dtos.BookDTOKt.toModel;

@Entity
@Table(name = "books")
public class Book extends BaseEntity{

    private String title = "";
    private String publisher = "";
    private String isbn = "";
    private Integer amount = 0;

    @ManyToMany
    @JoinTable(name = "visitor_book", joinColumns = @JoinColumn(name = "book_id"),
        inverseJoinColumns = @JoinColumn(name = "visitor_id"))
    private Set<Visitor> visitors = new HashSet<>();

    @ManyToMany
    @JoinTable(name = "author_book", joinColumns = @JoinColumn(name = "book_id"),
        inverseJoinColumns = @JoinColumn(name = "author_id"))
    private Set<Author> authors = new HashSet<>();

    public Book() {}

    public Book(String title, String publisher) {
        this.title = title;
        this.publisher = publisher;
    }

    public Book(Long id, String title, String publisher, String isbn, Integer amount,
                Set<Author> authors) {
        setId(id);
        this.title = title;
        this.publisher = publisher;
        this.isbn = isbn;
        this.amount = amount;
        this.visitors = visitors;
        this.authors = authors;
    }

    public String getTitle() {
        return title;
    }

    public void setTitle(String title) {
        this.title = title;
    }

    public String getPublisher() {
        return publisher;
    }

    public void setPublisher(String publisher) {
        this.publisher = publisher;
    }

    public String getIsbn() {
        return isbn;
    }

    public void setIsbn(String isbn) {
        this.isbn = isbn;
    }

    public Integer getAmount() {
        return amount;
    }

    public void setAmount(Integer amount) {
        this.amount = amount;
    }

    public Set<Visitor> getVisitors() {
        return visitors;
    }

    public void setVisitors(Set<Visitor> visitors) {
        this.visitors = visitors;
    }

    public Set<Author> getAuthors() {
        return authors;
    }

    public void setAuthors(Set<Author> authors) {
        this.authors = authors;
    }

    @Override
    public String toString() {
        return "Book{" +
                "title='" + title + '\'' +
                '}';
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;

        Book book = (Book) o;

        return this.getId().equals(book.getId());
    }

    public boolean rent(Visitor visitor){
        int amount = this.getAmount();
        boolean containsBook = visitor.containBook(this);
        if(amount > 0 && !containsBook){
            visitor.getBooks().add(this);
            this.getVisitors().add(visitor);
            this.setAmount(amount - 1);
            return true;
        }
        return false;
    }

    public boolean rent(VisitorDTO visitorDTO){
        Visitor visitor = toModel(visitorDTO);
        boolean isValid = rent(visitor);

        if(isValid) {
            visitorDTO.setBooks(visitor.books.stream().map(Book::toDto).collect(Collectors.toSet()));
            return true;
        }
        return false;


    }

    public boolean returnBook(Visitor visitor){
        if(visitor.containBook(this)) {
            boolean isBookRemoved = visitor.getBooks().removeIf(book -> book.equals(this));
            boolean isVisitorRemoved = this.getVisitors().removeIf(visit -> visit.equals(visitor));
            int amount = this.getAmount();
            this.setAmount(amount + 1);
            return isBookRemoved && isVisitorRemoved;
        }
        return false;
    }

    public boolean returnBook(VisitorDTO visitorDTO){
        Visitor visitor = toModel(visitorDTO);
        boolean isValid = returnBook(visitor);
        if(isValid){
            visitorDTO.setBooks(visitor.books.stream().map(Book::toDto).collect(Collectors.toSet()));
            return true;
        }
        return false;
    }

    public BookDTO toDto(){
        BookDTO bookDTO = new BookDTO(
            getId(), title, publisher, isbn, amount, authors
        );
        for(Visitor visitor: visitors){
            VisitorDTO visitorDTO = new VisitorDTO(
                    visitor.getId(),
                    visitor.getFirstName(),
                    visitor.getLastName(),
                    visitor.getPassword()
            );
            visitorDTO.getBooks().add(bookDTO);
            bookDTO.getVisitors().add(visitorDTO);
        }

        return bookDTO;
    }

    public static boolean isValid(Book book){
        if(book == null) return false;
        if(book.getTitle() != null && book.getPublisher() != null
                && book.getIsbn() != null){
            return !(book.getTitle().isEmpty() || book.getPublisher().isEmpty()
                    || book.getIsbn().isEmpty());
        }
        return false;
    }

    public static boolean isValid(BookDTO book){
        Book _book = toModel(book);
        return isValid(_book);
    }

    public static boolean isValidWithId(Book book){
        if(book != null){
            return Book.isValid(book) && book.getId() != null;
        } else return false;
    }
}

File: src/main/java/com/ottofernan/librarycrud/domain/models/Book.kt
//package com.ottofernan.librarycrud.domain.models
//
//import com.ottofernan.librarycrud.domain.dtos.BookDTO
//import com.ottofernan.librarycrud.domain.dtos.VisitorDTO
//import com.ottofernan.librarycrud.domain.dtos.toModel
//import java.util.stream.Collectors
//import javax.persistence.*
//
//@Entity
//@Table(name = "books")
//class Book(
//        var id: Long = 0,
//        var title: String = "",
//        var publisher: String = "",
//        var isbn: String = "",
//        var amount: Int = 0,
//
//        @ManyToMany
//        @JoinTable(name = "author_book",
//                joinColumns = [JoinColumn(name = "book_id", referencedColumnName = "id")],
//                inverseJoinColumns = [JoinColumn(name = "author_id", referencedColumnName = "id")])
//        var authors: MutableSet<Author> = HashSet()
//): BaseEntity(id) {
//
//    @ManyToMany
//    @JoinTable(name = "visitor_book",
//            joinColumns = [JoinColumn(name = "book_id", referencedColumnName = "id")],
//            inverseJoinColumns = [JoinColumn(name = "visitor_id", referencedColumnName = "id")])
//    var visitors: MutableSet<Visitor> = mutableSetOf()
//
//    override fun toString(): String {
//        return "Book{" +
//                "title='" + title + '\'' +
//                '}'
//    }
//
//    override fun equals(other: Any?): Boolean {
//        if (this === other) return true
//        if (other == null || javaClass != other.javaClass) return false
//        val book = other as Book
//        return id == book.id
//    }
//
//    fun rent(visitor: Visitor): Boolean {
//        val amount: Int = this.amount
//        val containsBook: Boolean = visitor.containBook(this)
//        if (amount > 0 && !containsBook) {
//            visitor.books.add(this)
//            this.visitors.add(visitor)
//            this.amount = amount - 1
//            return true
//        }
//        return false
//    }
//
//    fun rent(visitorDTO: VisitorDTO): Boolean {
//        val visitor = toModel(visitorDTO)
//        val isValid = rent(visitor)
//        if (isValid) {
//            visitorDTO.books = visitor.books.stream().map { obj: Book -> obj.toDto() }.collect(Collectors.toSet())
//            return true
//        }
//        return false
//    }
//
//    fun returnBook(visitor: Visitor): Boolean {
//        if (visitor.containBook(this)) {
//            val isBookRemoved = visitor.getBooks().removeIf { book: Book -> book == this }
//            val isVisitorRemoved: Boolean = this.visitors.removeIf { visit: Visitor -> visit == visitor }
//            val amount: Int = this.amount
//            this.amount = amount + 1
//            return isBookRemoved && isVisitorRemoved
//        }
//        return false
//    }
//
//    fun returnBook(visitorDTO: VisitorDTO): Boolean {
//        val visitor = toModel(visitorDTO)
//        val isValid = returnBook(visitor)
//        if (isValid) {
//            visitorDTO.books = visitor.books.stream().map { obj: Book -> obj.toDto() }.collect(Collectors.toSet())
//            return true
//        }
//        return false
//    }
//
//    fun toDto(): BookDTO? {
//        val bookDTO = BookDTO(
//                id, title, publisher, isbn, amount, authors
//        )
//        for (visitor in visitors) {
//            val visitorDTO = VisitorDTO(
//                    visitor.id,
//                    visitor.firstName,
//                    visitor.lastName,
//                    visitor.password
//            )
//            visitorDTO.books.add(bookDTO)
//            bookDTO.visitors.add(visitorDTO)
//        }
//        return bookDTO
//    }
//
//    companion object Validation {
//
//        fun isValid(book: Book?): Boolean {
//            if (book == null) return false
//            return !(book.title.isEmpty() || book.publisher.isEmpty()
//                    || book.isbn.isEmpty())
//        }
//
//        fun isValid(bookDTO: BookDTO?): Boolean {
//            val book = BookDTO.toModel(bookDTO)
//            return isValid(book)
//        }
//
//    }
//
//}